### ISSUE TITLE
Question regarding analyzing memory consumption

### ISSUE THREAD

**Issue by c-koell** [eclipse-openj9/openj9#15898]
We have a jakarta application on Open Liberty with openj9 jvm. Since a while we face a memory problem in the application but we are now at a point where we have no more idea to find the cause.

We have created a System dump to get more details about the memory consumption
```server dump MyServer --include=system```

The output of jdmpview displays following
```
JRE: 7468572800 bytes / 23468943 allocations
|  
+--VM: 6992801096 bytes / 23455300 allocations
|  |  
|  +--Classes: 228455016 bytes / 23901 allocations
|  |  |  
|  |  +--Shared Class Cache: 83886080 bytes / 1 allocations
|  |  |  
|  |  +--Other: 144568936 bytes / 23900 allocations
|  |  
|  +--Modules: 4131256 bytes / 19718 allocations
|  |  
|  +--Memory Manager (GC): 3289902288 bytes / 1902 allocations
|  |  |  
|  |  +--Java Heap: 3221286912 bytes / 1 allocations
|  |  |  
|  |  +--Other: 68615376 bytes / 1901 allocations
|  |  
|  +--Threads: 28420056 bytes / 349 allocations
|  |  |  
|  |  +--Java Stack: 1908824 bytes / 86 allocations
|  |  |  
|  |  +--Native Stack: 25821184 bytes / 87 allocations
|  |  |  
|  |  +--Other: 690048 bytes / 176 allocations
|  |  
|  +--Trace: 1506168 bytes / 4945 allocations
|  |  
|  +--JVMTI: 54800 bytes / 47 allocations
|  |  |  
|  |  +--JVMTI Allocate(): 184 bytes / 1 allocations
|  |  |  
|  |  +--Other: 54616 bytes / 46 allocations
|  |  
|  +--JNI: 4189816 bytes / 11392 allocations
|  |  
|  +--Port Library: 140896392 bytes / 86 allocations
|     |  
|     +--Unused <32bit allocation regions: 140883464 bytes / 1 allocations
|     |  
|     +--Other: 12928 bytes / 85 allocations
|  |  
|  +--Other: 3295245304 bytes / 23392960 allocations
|  
+--JIT: 466307656 bytes / 13397 allocations
|  |  
|  +--JIT Code Cache: 268435456 bytes / 1 allocations
|  |  
|  +--JIT Data Cache: 62916480 bytes / 30 allocations
|  |  
|  +--Other: 134955720 bytes / 13366 allocations
|  
+--Class Libraries: 9464048 bytes / 246 allocations
   |  
   +--VM Class Libraries: 9464048 bytes / 246 allocations
      |  
      +--sun.misc.Unsafe: 9250440 bytes / 198 allocations
         |  
         +--Direct Byte Buffers: 9154888 bytes / 143 allocations
         |  
         +--Other: 95552 bytes / 55 allocations
      |  
      +--Other: 213608 bytes / 48 allocationsNote: -Dcom.ibm.dbgmalloc=true was not found in the Java VM init options.
Memory allocated by some class library components will not have been recorded.
```
Here we see that Other Memory (3GB )is allocated outside of heap.
```+--Other: 3295245304 bytes / 23392960 allocations```

Is there a way or a tool that displays the root cause of the huge memory consumption outside of heap ?

Thanks for any help !

**Comment by pshipton** [comment]
All VM non-virtual memory allocations are tracked to the line of code doing the allocation. Run `jdmpview` on the core and run `!findallcallsites` to get the summary. Run `!j9help` to see a list of other possible commands. These are not documented outside of the help information. Unfortunately the callsites list is not correlated with the categories shown by `info memory`, although you can look at the line of code shown in the output to see which category was used for the allocation. Similarly you can search for the category in the code to find allocations that use the category. "Other" under another header such as "VM" means the memory was allocated directly in the "VM" category. In this case the constant in the code is OMRMEM_CATEGORY_VM as seen in omrmemcategories.h, but see also j9memcategories.h and j9memcategories.c.

You can also get a summary of allocations for a running application by using the command line option `-Xcheck:memory:quick,noscan,callsite=99999`. This outputs a summary of allocations every 99999th allocation. You can change this value to get more or less output.

**Comment by pshipton** [comment]
FYI https://github.com/ibmruntimes/Semeru-Runtimes/issues/32 which is fixed in the latest releases (0.33)

**Comment by c-koell** [comment]
Thanks very much for the info. We see also a lot of memory allocated by extendedMessageNPE.cpp
```
  total alloc  | largest
blocks | bytes | bytes | callsite
-------+-------+-------+-------+-------+-------+-------+-------+-------+-------
   5233 201219672 2097152 c:\workspace\openjdk-build\workspace\build\src\openj9\runtime\vm\segment.c:253
1100591 123267600    1520 c:\workspace\openjdk-build\workspace\build\src\openj9\runtime\vm\extendedMessageNPE.cpp:609
1100591 123267600    1520 c:\workspace\openjdk-build\workspace\build\src\openj9\runtime\vm\extendedMessageNPE.cpp:608
1100591 79243392     912 c:\workspace\openjdk-build\workspace\build\src\openj9\runtime\vm\extendedMessageNPE.cpp:619
1100591 79242632     152 c:\workspace\openjdk-build\workspace\build\src\openj9\runtime\vm\extendedMessageNPE.cpp:604
   7115 63039248   41240 c:\workspace\openjdk-build\workspace\build\src\openj9\runtime\vm\segment.c:248
1100591 17609496      56 c:\workspace\openjdk-build\workspace\build\src\openj9\runtime\vm\extendedMessageNPE.cpp:637
   4966 9972232   49456 c:\workspace\openjdk-build\workspace\build\src\openj9\runtime\compiler\runtime\MethodMetaData.c:189
    162 9163085  327696 c:\workspace\openjdk-build\workspace\build\src\openj9\runtime\jcl\common\unsafe_mem.c:241
      5 6543760 1308752 c:\workspace\openjdk-build\workspace\build\src\omr\gc\base\WorkPackets.cpp:179
    900 3496928  300584 c:\workspace\openjdk-build\workspace\build\src\openj9\runtime\gc_base\StringTable.cpp:86
   1280 2623368    2072 c:\workspace\openjdk-build\workspace\build\src\openj9\runtime\vm\segment.c:442
```
We will try the newest version 17.0.4.1+1
